modules/loki/main.tf

```
@@ -1,11 +1,162 @@
+            dynamodb = {
+              dynamodb_url = "dynamodb://${data.aws_region.loki.name}"
+              metrics = {
+                url: "http://prometheus-server.kube-system.svc.cluster.local:9090"
+              }
+            }
+          }
+        }
+        table_manager = {
+          retention_deletes_enabled = true
+          retention_period = "720h"
+          index_tables_provisioning = {
+            provisioned_write_throughput = 30
+            provisioned_read_throughput = 3
+            write_scale = {
+              enabled = true
+              min_capacity = 1
+              role_arn = module.iam.this_iam_role_arn
+            }
+            read_scale = {
+              enabled = true
+              min_capacity = 1
+              role_arn = module.iam.this_iam_role_arn
+            }
+          }
+        }
+      }
+    }
+  }
 }
 
+data aws_caller_identity "loki" {}
+data aws_region "loki" {}
+
+module "iam" {
+  source = "terraform-aws-modules/iam/aws//modules/iam-assumable-role-with-oidc"
+
+  create_role                   = var.enable
+  role_name                     = "${var.cluster_name}-loki-irsa"
+  provider_url                  = local.provider_url
+  oidc_fully_qualified_subjects = ["system:serviceaccount:${local.namespace}:loki"]
+
+  tags = var.tags
+}
+
+data "aws_iam_policy_document" "loki" {
+  statement {
+    actions = [
+      "s3:ListBucket",
+      "s3:PutObject",
+      "s3:GetObject"
+    ]
+    resources = ["arn:aws:s3:::${local.bucket_name}", "arn:aws:s3:::${local.bucket_name}/*"]
+  }
+
+  statement {
+    actions   = ["dynamodb:ListTables"]
+    resources = ["*"]
+  }
+
+  statement {
+    actions = [
+      "dynamodb:BatchGetItem",
+      "dynamodb:BatchWriteItem",
+      "dynamodb:DeleteItem",
+      "dynamodb:DescribeTable",
+      "dynamodb:GetItem",
+      "dynamodb:ListTagsOfResource",
+      "dynamodb:PutItem",
+      "dynamodb:Query",
+      "dynamodb:TagResource",
+      "dynamodb:UntagResource",
+      "dynamodb:UpdateItem",
+      "dynamodb:UpdateTable",
+      "dynamodb:CreateTable",
+      "dynamodb:DeleteTable"
+    ]
+
+    resources = ["arn:aws:dynamodb:${data.aws_region.loki.name}:${data.aws_caller_identity.loki.account_id}:table/${local.dynamodb_table}*"]
+  }
+
+  statement {
+    actions = [
+      "application-autoscaling:DescribeScalableTargets",
+      "application-autoscaling:DescribeScalingPolicies",
+      "application-autoscaling:RegisterScalableTarget",
+      "application-autoscaling:DeregisterScalableTarget",
+      "application-autoscaling:PutScalingPolicy",
+      "application-autoscaling:DeleteScalingPolicy"
+    ]
+    resources = ["*"]
+  }
+
+  statement {
+    actions = [
+      "iam:GetRole",
+      "iam:PassRole"
+    ]
+    resources = [
+      module.iam.this_iam_role_arn
+    ]
+  }
+
+
+}
+
+resource "aws_iam_role_policy" "loki" {
+  count = var.enable ? 1 : 0
+
+  name = "LokiStorage"
```

This will cause a conflict if more instances are used in the same account. Should use `name_prefix` instead.

```suggestion
  name_prefix = "LokiStorage_"
```